Code/GraphMol/DistGeomHelpers/Embedder.cpp

```diff
@@ -721,42 +756,91 @@ bool embedPoints(RDGeom::PointPtrVect *positions, detail::EmbedArgs eargs,
+#endif
+        embedParams.failures[EmbedFailureCauses::INITIAL_COORDS]++;
+      }
+    } else {
       gotCoords =
           EmbeddingOps::firstMinimization(positions, eargs, embedParams);
-      if (gotCoords) {
+      if (!gotCoords) {
+        if (embedParams.trackFailures) {
+#ifdef RDK_BUILD_THREADSAFE_SSS
+          std::lock_guard<std::mutex> lock(GetFailMutex());
+#endif
+          embedParams.failures[EmbedFailureCauses::FIRST_MINIMIZATION]++;
+        }
+      } else {
         gotCoords = EmbeddingOps::checkTetrahedralCenters(positions, eargs,
                                                           embedParams);
+        if (!gotCoords) {
+          if (embedParams.trackFailures) {
+#ifdef RDK_BUILD_THREADSAFE_SSS
+            std::lock_guard<std::mutex> lock(GetFailMutex());
+#endif
+            embedParams
+                .failures[EmbedFailureCauses::CHECK_TETRAHEDRAL_CENTERS]++;
+          }
+        }
       }
 
       // Check if any of our chiral centers are badly out of whack.
       if (gotCoords && embedParams.enforceChirality &&
           eargs.chiralCenters->size() > 0) {
         gotCoords =
             EmbeddingOps::checkChiralCenters(positions, eargs, embedParams);
+        if (!gotCoords) {
+          if (embedParams.trackFailures) {
+#ifdef RDK_BUILD_THREADSAFE_SSS
+            std::lock_guard<std::mutex> lock(GetFailMutex());
+#endif
+            embedParams.failures[EmbedFailureCauses::CHECK_CHIRAL_CENTERS]++;
+          }
+        }
       }
       // redo the minimization if we have a chiral center
       // or have started from random coords.
       if (gotCoords &&
           (eargs.chiralCenters->size() > 0 || embedParams.useRandomCoords)) {
         gotCoords = EmbeddingOps::minimizeFourthDimension(positions, eargs,
                                                           embedParams);
+        if (!gotCoords) {
+          if (embedParams.trackFailures) {
+#ifdef RDK_BUILD_THREADSAFE_SSS
+            std::lock_guard<std::mutex> lock(GetFailMutex());
+#endif
+            embedParams
+                .failures[EmbedFailureCauses::MINIMIZE_FOURTH_DIMENSION]++;
+          }
+        }
       }
 
       // (ET)(K)DG
       if (gotCoords && (embedParams.useExpTorsionAnglePrefs ||
                         embedParams.useBasicKnowledge)) {
         gotCoords = EmbeddingOps::minimizeWithExpTorsions(*positions, eargs,
                                                           embedParams);
+        if (!gotCoords) {
+          if (embedParams.trackFailures) {
+#ifdef RDK_BUILD_THREADSAFE_SSS
+            std::lock_guard<std::mutex> lock(GetFailMutex());
+#endif
+            embedParams.failures[EmbedFailureCauses::ETK_MINIMIZATION]++;
+          }
+        }
       }
       // test if chirality is correct
       if (embedParams.enforceChirality && gotCoords &&
           (eargs.chiralCenters->size() > 0)) {
         gotCoords =
             EmbeddingOps::finalChiralChecks(positions, eargs, embedParams);
```

Comment:
Maybe add a comment that the tracking of the fails for this step is done within finalChiralChecks()